src/python_framework_v02/nhd_io.py

```diff
@@ -55,9 +47,202 @@ def replace_downstreams(data, downstream_col, terminal_code):
     return new_data
 
 
-def read_waterbody_df(parm_file, lake_id_mask=None):
-    df1 = xr.open_dataset(parm_file).to_dataframe()
-    df1 = df1.set_index("lake_id").sort_index()
+def read_waterbody_df(waterbody_parameters, waterbodies_values, wbtype="level_pool"):
+    """
+    General waterbody dataframe reader. At present, only level-pool
+    capability exists. 
+    """
+    if wbtype == "level_pool":
+        wb_params = waterbody_parameters[wbtype]
+        return read_level_pool_waterbody_df(
+            wb_params["level_pool_waterbody_parameter_file_path"],
+            wb_params["level_pool_waterbody_id"],
+            waterbodies_values[wbtype],
+        )
+
+
+def read_level_pool_waterbody_df(
+    parm_file, lake_index_field="lake_id", lake_id_mask=None
+):
+    """
+    Reads LAKEPARM file and prepares a dataframe, filtered 
+    to the relevant reservoirs, to provide the parameters
+    for level-pool reservoir computation.
+
+    Completely replaces the read_waterbody_df function from prior versions
+    of the v02 routing code.
+    """
+
+    with xr.open_dataset(parm_file) as ds:
+        df1 = ds.to_dataframe()
+    df1 = df1.set_index(lake_index_field).sort_index(axis="index")
     if lake_id_mask is None:
         return df1
     return df1.loc[lake_id_mask]
+
+
+def get_ql_from_csv(qlat_input_file, index_col=0):
+    """
+    qlat_input_file: comma delimted file with header giving timesteps, rows for each segment
+    index_col = 0: column/field in the input file with the segment/link id
+    """
+    ql = pd.read_csv(qlat_input_file, index_col=index_col)
+    ql.index = ql.index.astype(int)
+    ql.columns = ql.columns.astype(int)
+    ql = ql.sort_index(axis="index")
+    return ql.astype("float32")
+
+
+def read_qlat(path):
+    """
+    retained for backwards compatibility with early v02 files
+    """
+    return get_ql_from_csv(path)
+
+
+def get_ql_from_wrf_hydro(qlat_files, index_col="station_id", value_col="q_lateral"):
+    """
+    qlat_files: globbed list of CHRTOUT files containing desired lateral inflows
+    index_col: column/field in the CHRTOUT files with the segment/link id
+    value_col: column/field in the CHRTOUT files with the lateral inflow value
+
+    In general the CHRTOUT files contain one value per time step. At present, there is
+    no capability for handling non-uniform timesteps in the qlaterals.
+
+    The qlateral may also be input using comma delimited file -- see
+    `get_ql_from_csv`
+    """
+
+    li = []
+
+    for filename in qlat_files:
+        ds = xr.open_dataset(filename)
```

Comment:
There should be a `ds.close()` somewhere here.

Reply:
Converted to `with` style